pkg/server/api/agent/v1/service.go

```diff
@@ -436,7 +540,183 @@ func (s *Service) createJoinTokenRegistrationEntry(ctx context.Context, token st
 	return nil
 }
 
-func applyMask(a *types.Agent, mask *types.AgentMask) { //nolint: unused,deadcode
+func (s *Service) updateAttestedNode(ctx context.Context, req *datastore.UpdateAttestedNodeRequest, log logrus.FieldLogger) error {
+	_, err := s.ds.UpdateAttestedNode(ctx, req)
+	switch status.Code(err) {
+	case codes.OK:
+		return nil
+	case codes.NotFound:
+		log.WithError(err).Error("Agent not found")
+		return status.Errorf(codes.NotFound, "agent not found: %v", err)
+	default:
+		log.WithError(err).Error("Failed to update agent")
+		return status.Errorf(codes.Internal, "failed to update agent: %v", err)
+	}
+}
+
+func (s *Service) signSvid(ctx context.Context, agentID *spiffeid.ID, csr []byte, log logrus.FieldLogger) ([]*x509.Certificate, error) {
+	parsedCsr, err := x509.ParseCertificateRequest(csr)
+	if err != nil {
+		log.WithError(err).Error("Invalid argument: failed to parse CSR")
+		return nil, status.Errorf(codes.InvalidArgument, "failed to parse CSR: %v", err)
+	}
+
+	// Sign a new X509 SVID
+	x509Svid, err := s.ca.SignX509SVID(ctx, ca.X509SVIDParams{
+		SpiffeID:  agentID.String(),
+		PublicKey: parsedCsr.PublicKey,
+	})
+	if err != nil {
+		log.WithError(err).Error("Failed to sign X509 SVID")
+		return nil, status.Errorf(codes.Internal, "failed to sign X509 SVID: %v", err)
+	}
+
+	return x509Svid, nil
+}
+
+func (s *Service) getSelectorsFromAgentID(ctx context.Context, agentID string) ([]*types.Selector, error) {
+	resp, err := s.ds.GetNodeSelectors(ctx, &datastore.GetNodeSelectorsRequest{
+		SpiffeId: agentID,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to get node selectors: %v", err)
+	}
+
+	return api.NodeSelectorsToProto(resp.Selectors)
+}
+
+func (s *Service) attestJoinToken(ctx context.Context, token string) (*nodeattestor.AttestResponse, error) {
+	log := rpccontext.Logger(ctx).WithField(telemetry.NodeAttestorType, "join_token")
+
+	resp, err := s.ds.FetchJoinToken(ctx, &datastore.FetchJoinTokenRequest{
+		Token: token,
+	})
+	switch {
+	case err != nil:
+		log.WithError(err).Error("Failed to fetch join token")
+		return nil, status.Error(codes.Internal, "failed to fetch join token")
+	case resp.JoinToken == nil:
+		log.Error("Join token does not exist or has already been used")
+		return nil, status.Error(codes.InvalidArgument, "join token does not exist or has already been used")
```

Comment:
This fails attestation. To keep consistency with other errors that start with "failed to attest", I suggest to change this to "failed to attest: join token does not exist or has already been used".

nit: some start with "failed to attest," and others start with "failed to attest;", we may change all to start with "failed to attest:"